File: research/experiments/sweep/runs/opt-large-go-sonnet45/finite_diff.go
package optimization

import "math"

const machineEpsilon = 2.220446049250313e-16

// ForwardDiffGradient computes the gradient using forward differences
func ForwardDiffGradient(f func([]float64) float64, x []float64) []float64 {
	n := len(x)
	grad := make([]float64, n)
	fx := f(x)
	xPerturb := Clone(x)

	sqrtEps := math.Sqrt(machineEpsilon)

	for i := 0; i < n; i++ {
		h := sqrtEps * math.Max(math.Abs(x[i]), 1.0)
		xPerturb[i] = x[i] + h
		fxPlusH := f(xPerturb)
		grad[i] = (fxPlusH - fx) / h
		xPerturb[i] = x[i] // restore
	}

	return grad
}

// CentralDiffGradient computes the gradient using central differences
func CentralDiffGradient(f func([]float64) float64, x []float64) []float64 {
	n := len(x)
	grad := make([]float64, n)
	xPerturb := Clone(x)

	cbrtEps := math.Pow(machineEpsilon, 1.0/3.0)

	for i := 0; i < n; i++ {
		h := cbrtEps * math.Max(math.Abs(x[i]), 1.0)
		xPerturb[i] = x[i] + h
		fxPlusH := f(xPerturb)
		xPerturb[i] = x[i] - h
		fxMinusH := f(xPerturb)
		grad[i] = (fxPlusH - fxMinusH) / (2.0 * h)
		xPerturb[i] = x[i] // restore
	}

	return grad
}

// MakeGradient returns a gradient function using the specified method
func MakeGradient(f func([]float64) float64, method string) func([]float64) []float64 {
	if method == "central" {
		return func(x []float64) []float64 {
			return CentralDiffGradient(f, x)
		}
	}
	// Default to forward
	return func(x []float64) []float64 {
		return ForwardDiffGradient(f, x)
	}
}

File: research/experiments/sweep/runs/opt-medium-csharp-opus45/src/FiniteDiff.cs
using System;

namespace Optimization
{
    public static class FiniteDiff
    {
        // C# double.Epsilon is the smallest positive subnormal (~5e-324), NOT machine epsilon.
        // Machine epsilon for IEEE 754 double is 2^-52 ~ 2.22e-16.
        private const double MachineEpsilon = 2.220446049250313e-16;
        private static readonly double SqrtEps = Math.Sqrt(MachineEpsilon); // ~1.49e-8
        private static readonly double CbrtEps = Math.Cbrt(MachineEpsilon); // ~6.06e-6

        /// <summary>
        /// Forward difference gradient: g_i = (f(x + h*e_i) - f(x)) / h
        /// </summary>
        public static double[] ForwardDiffGradient(Func<double[], double> f, double[] x)
        {
            int n = x.Length;
            double fx = f(x);
            var grad = new double[n];

            for (int i = 0; i < n; i++)
            {
                double h = SqrtEps * Math.Max(Math.Abs(x[i]), 1.0);
                var xPerturbed = (double[])x.Clone();
                xPerturbed[i] += h;
                grad[i] = (f(xPerturbed) - fx) / h;
            }

            return grad;
        }

        /// <summary>
        /// Central difference gradient: g_i = (f(x + h*e_i) - f(x - h*e_i)) / (2h)
        /// </summary>
        public static double[] CentralDiffGradient(Func<double[], double> f, double[] x)
        {
            int n = x.Length;
            var grad = new double[n];

            for (int i = 0; i < n; i++)
            {
                double h = CbrtEps * Math.Max(Math.Abs(x[i]), 1.0);
                var xPlus = (double[])x.Clone();
                var xMinus = (double[])x.Clone();
                xPlus[i] += h;
                xMinus[i] -= h;
                grad[i] = (f(xPlus) - f(xMinus)) / (2 * h);
            }

            return grad;
        }

        /// <summary>
        /// Factory: returns a gradient function using the specified method.
        /// </summary>
        public static Func<double[], double[]> MakeGradient(
            Func<double[], double> f, string method = "forward")
        {
            if (method == "central")
                return x => CentralDiffGradient(f, x);
            return x => ForwardDiffGradient(f, x);
        }
    }
}
